Bull Missing Discard Method

If you know how to fix the issue, make a pull request instead.

- [X] I tried using the `@types/bull` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [ ] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @bgrieder, @JProgrammer, @marshall007, @weeco

If you do not mention the authors the issue will be ignored.

Im Trying to use discard method of a job how cited here: https://github.com/OptimalBits/bull/blob/master/REFERENCE.md#jobdiscard
but index.d.ts does not have this method.

src/queues/pending-trip-request.ts(52,11): error TS2339: Property 'discard' does not exist on type 'Job'.
npm ERR! code ELIFECYCLE
npm ERR! errno 2

--- a/types/bull/bull-tests.tsx
+++ b/types/bull/bull-tests.tsx
@@ -158,9 +158,21 @@ myQueue.on('active', (job: Queue.Job) => {
     job.moveToCompleted();
     job.moveToCompleted('done');
     job.moveToCompleted('done', true);
+    job.moveToCompleted('done', true).then(val => {
+        if (val) {
+            const nextJobData: any = val[0];
+            const nextJobId: Queue.JobId = val[1];
+        }
+    });
 
     job.moveToFailed({ message: "Call to external service failed!" }, true);
     job.moveToFailed(new Error('test error'), true);
+    job.moveToFailed(new Error('test error'), true).then(val => {
+        if (val) {
+            const nextJobData: any = val[0];
+            const nextJobId: Queue.JobId = val[1];
+        }
+    });
 
     job.discard();
 });

--- a/types/bull/bull-tests.tsx
+++ b/types/bull/bull-tests.tsx
@@ -125,6 +125,7 @@ videoQueue.add({ video: 'http://example.com/video1.mov' }, { jobId: 1 })
 pdfQueue
 .on('error', (err: Error) => undefined)
 .on('active', (job: Queue.Job, jobPromise: Queue.JobPromise) => jobPromise.cancel())
+.on('waiting', (jobId: Queue.JobId) => undefined)
 .on('active', (job: Queue.Job) => undefined)
 .on('stalled', (job: Queue.Job) => undefined)
 .on('progress', (job: Queue.Job) => undefined)
@@ -132,4 +133,41 @@ pdfQueue
 .on('failed', (job: Queue.Job) => undefined)
 .on('paused', () => undefined)
 .on('resumed', () => undefined)
-.on('cleaned', (jobs: Queue.Job[], status: Queue.JobStatus) => undefined);
+.on('cleaned', (jobs: Queue.Job[], status: Queue.JobStatus) => undefined)
+.on('drained', () => undefined)
+.on('removed', (job: Queue.Job) => undefined);
+
+// test different process methods
+
+const profileQueue = new Queue('profile');
+// Max concurrency for requestProfile is 100
+profileQueue.process('requestProfile', 100, () => {});
+profileQueue.process(100, () => {});
+
+// other tests
+const myQueue = new Queue('myQueue', {
+    settings: {
+        drainDelay: 5
+    },
+    defaultJobOptions: {
+        stackTraceLimit: 1,
+    }
+});
+
+myQueue.on('active', (job: Queue.Job) => {
+    job.moveToCompleted();
+    job.moveToCompleted('done');
+    job.moveToCompleted('done', true);
+
+    job.moveToFailed({ message: "Call to external service failed!" }, true);
+    job.moveToFailed(new Error('test error'), true);
+
+    job.discard();
+});
+
+// test all constructor options:
+
+new Queue('profile');
+new Queue('profile', 'url');
+new Queue('profile', { prefix: 'test' });
+new Queue('profile', 'url', { prefix: 'test' });

--- a/types/bull/index.d.ts
+++ b/types/bull/index.d.ts
@@ -164,13 +164,13 @@ declare namespace Bull {
      * Moves a job to the `completed` queue. Pulls a job from 'waiting' to 'active'
      * and returns a tuple containing the next jobs data and id. If no job is in the `waiting` queue, returns null.
      */
-    moveToCompleted(returnValue?: string, ignoreLock?: boolean): Promise<string[] | null>;
+    moveToCompleted(returnValue?: string, ignoreLock?: boolean): Promise<[any, JobId] | null>;
 
     /**
      * Moves a job to the `failed` queue. Pulls a job from 'waiting' to 'active'
      * and returns a tuple containing the next jobs data and id. If no job is in the `waiting` queue, returns null.
      */
-    moveToFailed(errorInfo: { message: string; }, ignoreLock?: boolean): Promise<void>;
+    moveToFailed(errorInfo: { message: string; }, ignoreLock?: boolean): Promise<[any, JobId] | null>;
 
     /**
      * Promotes a job that is currently "delayed" to the "waiting" state and executed as soon as possible.

--- a/types/bull/index.d.ts
+++ b/types/bull/index.d.ts
@@ -1,4 +1,4 @@
-// Type definitions for bull 3.3
+// Type definitions for bull 3.4
 // Project: https://github.com/OptimalBits/bull
 // Definitions by: Bruno Grieder <https://github.com/bgrieder>
 //                 Cameron Crothers <https://github.com/JProgrammer>
@@ -10,6 +10,7 @@
 //                 Bond Akinmade <https://github.com/bondz>
 //                 Wuha Team <https://github.com/wuha-team>
 //                 Alec Brunelle <https://github.com/aleccool213>
+//                 Dan Manastireanu <https://github.com/danmana>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 2.8
 
@@ -23,9 +24,9 @@ import * as Promise from "bluebird";
  */
 declare const Bull: {
   (queueName: string, opts?: Bull.QueueOptions): Bull.Queue;
-  (queueName: string, url?: string): Bull.Queue; // tslint:disable-line unified-signatures
+  (queueName: string, url: string, opts?: Bull.QueueOptions): Bull.Queue; // tslint:disable-line unified-signatures
   new (queueName: string, opts?: Bull.QueueOptions): Bull.Queue;
-  new (queueName: string, url?: string): Bull.Queue; // tslint:disable-line unified-signatures
+  new (queueName: string, url: string, opts?: Bull.QueueOptions): Bull.Queue; // tslint:disable-line unified-signatures
 };
 
 declare namespace Bull {
@@ -92,6 +93,12 @@ declare namespace Bull {
     backoffStrategies?: {
       [key: string]: (attemptsMade: number, err: typeof Error) => number;
     };
+
+    /**
+     * A timeout for when the queue is in `drained` state (empty waiting for jobs).
+     * It is used when calling `queue.getNextJob()`, which will pass it to `.brpoplpush` on the Redis client.
+     */
+    drainDelay?: number;
   }
 
   type DoneCallback = (error?: Error | null, value?: any) => void;
@@ -142,11 +149,28 @@ declare namespace Bull {
     retry(): Promise<void>;
 
     /**
+     * Ensure this job is never ran again even if attemptsMade is less than job.attempts.
+     */
+    discard(): Promise<void>;
+
+    /**
      * Returns a promise that resolves to the returned data when the job has been finished.
      * TODO: Add a watchdog to check if the job has finished periodically.
      * since pubsub does not give any guarantees.
      */
     finished(): Promise<any>;
+
+    /**
+     * Moves a job to the `completed` queue. Pulls a job from 'waiting' to 'active'
+     * and returns a tuple containing the next jobs data and id. If no job is in the `waiting` queue, returns null.
+     */
+    moveToCompleted(returnValue?: string, ignoreLock?: boolean): Promise<string[] | null>;
+
+    /**
+     * Moves a job to the `failed` queue. Pulls a job from 'waiting' to 'active'
+     * and returns a tuple containing the next jobs data and id. If no job is in the `waiting` queue, returns null.
+     */
+    moveToFailed(errorInfo: { message: string; }, ignoreLock?: boolean): Promise<void>;
 
     /**
      * Promotes a job that is currently "delayed" to the "waiting" state and executed as soon as possible.
@@ -205,6 +229,11 @@ declare namespace Bull {
      * Cron pattern specifying when the job should execute
      */
     cron: string;
+
+    /**
+     * Start date when the repeat job should start repeating (only with cron).
+     */
+    startDate?: Date | string | number;
   }
 
   interface EveryRepeatOptions extends RepeatOptions {
@@ -273,6 +302,11 @@ declare namespace Bull {
      * Default behavior is to keep the job in the completed set.
      */
     removeOnFail?: boolean;
+
+    /**
+     * Limits the amount of stack trace lines that will be recorded in the stacktrace.
+     */
+    stackTraceLimit?: number;
   }
 
   interface JobCounts {
@@ -619,6 +653,11 @@ declare namespace Bull {
     on(event: 'error', callback: ErrorEventCallback): this;
 
     /**
+     * A Job is waiting to be processed as soon as a worker is idling.
+     */
+    on(event: 'waiting', callback: WaitingEventCallback): this;
+
+    /**
      * A job has started. You can use `jobPromise.cancel()` to abort it
      */
     on(event: 'active', callback: ActiveEventCallback<T>): this;
@@ -655,12 +694,23 @@ declare namespace Bull {
     on(event: 'resumed', callback: EventCallback): this; // tslint:disable-line unified-signatures
 
     /**
+     * A job successfully removed.
+     */
+    on(event: 'removed', callback: RemovedEventCallback<T>): this;
+
+    /**
      * Old jobs have been cleaned from the queue.
      * `jobs` is an array of jobs that were removed, and `type` is the type of those jobs.
      *
      * @see Queue#clean() for details
      */
     on(event: 'cleaned', callback: CleanedEventCallback<T>): this;
+
+    /**
+     * Emitted every time the queue has processed all the waiting jobs
+     * (even if there can be some delayed jobs not yet processed)
+     */
+    on(event: 'drained', callback: EventCallback): this; // tslint:disable-line unified-signatures
   }
 
   type EventCallback = () => void;
@@ -685,6 +735,10 @@ declare namespace Bull {
   type FailedEventCallback<T = any> = (job: Job<T>, error: Error) => void;
 
   type CleanedEventCallback<T = any> = (jobs: Array<Job<T>>, status: JobStatus) => void;
+
+  type RemovedEventCallback<T = any> = (job: Job<T>) => void;
+
+  type WaitingEventCallback = (jobId: JobId) => void;
 }
 
 export = Bull;

